oscarlab/graphene
# [Pal/Linux-SGX] Allow enclave sizes greater than 32GB

## Description of the changes 

Previously, Graphene had unnecessarily restrictive logic to handle PIE and non-PIE executables.

Non-PIE executables must be accommodated in an enclave that spans address 0x400000, so enclaves for non-PIE
must have base address of 0x0 and heap base of not 0x0. PIE executables have no restrictions on available addresses, so enclaves for PIE may have any base address (but aligned to enclave size due to SGX HW quirk) and heap base can start at enclave base.

Previously, for PIE executables Graphene chose enclave base = `ENCLAVE_HIGH_ADDRESS` = 32GB; this led to impossibility of creating enclaves of sizes greater than 32GB (due to the aforementioned SGX HW quirk).

This PR refactors all this logic and removes the hard-coded `ENCLAVE_HIGH_ADDRESS`. Instead, PIE executables have enclave base address same as enclave size (i.e., a 64GB enclave starts at address 64GB).

Fixes #1856.

## How to test this PR? 

No new tests. Try 64GB enclaves. Also, add `sgx.static_address = 0` and `sgx.static_address = 1` to say Redis or Memcached and observe debug logs.


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/oscarlab/graphene/1863)
